### PR TITLE
feat(query): implement copy from ipfs

### LINKED ISSUE #7659: Feature: Allow COPY FROM IPFS
**Summary**

OpenDAL already has IPFS support. Databend can implement copy from ipfs now:

```sql
COPY INTO my_table FROM 'ipfs://QmHash' FILE_FORMAT=( TYPE = 'JSON' );
```

### PR DESCRIPTION
**Pull request by ClSlaid**
Signed-off-by: ClSlaid <cailue@bupt.edu.cn>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This PR provides a feature of copy files from IPFS.

Fixes #7659

### CODE REVIEW

File: src/common/storage/Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,7 @@ edition = "2021"
 
 [features]
 storage-hdfs = ["opendal/services-hdfs"]
+storage-ipfs = ["opendal/services-ipfs"]
```
Comment (Xuanwo) [inline src/common/storage/Cargo.toml:11]: Please always enable this feature. Databend already depends on `prost`.

File: src/common/storage/src/location.rs [hunk 2]
```diff
@@ -95,6 +95,15 @@ pub fn parse_uri_location(l: &UriLocation) -> Result<(StorageParams, String)> {
                 .to_string(),
             root: root.to_string(),
         }),
+        #[cfg(feature = "storage-ipfs")]
+        Scheme::Ipfs => {
+            use crate::config::StorageIpfsConfig;
+            use crate::config::STORAGE_IPFS_DEFAULT_ENDPOINT;
+            StorageParams::Ipfs(StorageIpfsConfig {
+                endpoint_url: STORAGE_IPFS_DEFAULT_ENDPOINT.to_string(),
+                root: "/ipfs/".to_string(),
```
Comment (Xuanwo) [inline src/common/storage/src/location.rs:104]: Root should be `format("/ipfs/{root}/")`?

**Comment by ClSlaid** [comment]
NOw suppOrts directories and FIles.

1. directories
图片
图片
**remember the forward slash -_- in the link.**
2. files
图片
**No forward slash in link**

File: src/common/storage/src/config.rs [hunk 3]
```diff
@@ -254,6 +259,14 @@ pub struct StorageHttpConfig {
     pub paths: Vec<String>,
 }
 
+pub const STORAGE_IPFS_DEFAULT_ENDPOINT: &str = "https://ipfs.filebase.io";
```
Comment (Xuanwo) [inline src/common/storage/src/config.rs:262]: Please use `https://ipfs.io` here.

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 4]
```diff
@@ -82,6 +85,7 @@ impl CopyInterpreterV2 {
                         vec![]
                     }
                 } else {
+                    tracing::info!("shit");
```
Comment (Xuanwo) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:88]: ?

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 5]
```diff
@@ -91,10 +95,13 @@ impl CopyInterpreterV2 {
 
                     // TODO: we could rewrite into try_collect.
                     let mut objects = op.batch().walk_top_down(path)?;
+                    tracing::info!("common here we go top down");
```
Comment (Xuanwo) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:98]: Please don't print meaningless logs to users.

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 6]
```diff
@@ -303,6 +310,10 @@ impl Interpreter for CopyInterpreterV2 {
 
                 // Pattern match check.
                 let pattern = &pattern;
+                tracing::info!(
```
Comment (Xuanwo) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:313]: ?

**Comment by Xuanwo** [comment]
Please add a stateful test called: `00_copy/00_0002_copy_from_ipfs_location.sh`

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../../../shell_env.sh
+TABLE=ontime200
+FILE=ontime_200
+QMHASH=Qmei4dyyPazUy24cfCAtmumC1Ff1VLLiqjzF1zCXXVtvSk
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:7]: Where is this hash from?

Comment (Xuanwo) [reply tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:7]: I'm guessing this hash is not reliable.

Please use this https://ipfs.filebase.io/ipfs/QmPpCt1aYGb9JWJRmXRUnmJtVgeFFTJGzWFYEEX7bo9zGJ for test.

Comment (ClSlaid) [reply tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:7]: 100000 lines of data are too large and will make the CI slow, so I uploaded the testing files on my computer. But...well seems like I know nothing about IPFS...

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+QMHASH=QmPpCt1aYGb9JWJRmXRUnmJtVgeFFTJGzWFYEEX7bo9zGJ
+
+echo "drop table if exists ${TABLE};" | $MYSQL_CLIENT_CONNECT
+
+## Create table
+cat $CURDIR/../ddl/ontime.sql | sed "s/ontime/$TABLE/g" | $MYSQL_CLIENT_CONNECT
+
+copy_from_location_cases=(
+  # copy csv
+  "copy into $TABLE from 'ipfs://$QMHASH/$FILE.csv' FILE_FORMAT = (type = 'CSV' field_delimiter = ','  record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:16]:
```suggestion
  "copy into $TABLE from 'ipfs://$QMHASH/ontime.csv' FILE_FORMAT = (type = 'CSV' field_delimiter = ','  record_delimiter = '\n' skip_header = 1)"
```

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  # copy gzip csv
+  "copy into $TABLE from 'ipfs://$QMHASH/$FILE.csv.gz' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'gzip'  record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:18]:
```suggestion
  "copy into $TABLE from 'ipfs://$QMHASH/ontime.csv.gz' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'gzip'  record_delimiter = '\n' skip_header = 1)"
```

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  # copy zstd csv
+  "copy into $TABLE from 'ipfs://$QMHASH/$FILE.csv.zstd' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'zstd'  record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:20]:
```suggestion
  "copy into $TABLE from 'ipfs://$QMHASH/ontime.csv.zst' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'zstd'  record_delimiter = '\n' skip_header = 1)"
```

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  # copy bz2 csv
+  "copy into $TABLE from 'ipfs://$QMHASH/$FILE.csv.bz2' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'bz2'  record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:22]:
```suggestion
  "copy into $TABLE from 'ipfs://$QMHASH/ontime.csv.bz2' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'bz2'  record_delimiter = '\n' skip_header = 1)"
```

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  # copy xz csv
+  "copy into $TABLE from 'ipfs://$QMHASH/$FILE.csv.xz' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'xz'  record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:24]: We don't have this file. Remove this case.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  # copy file
+  "copy into $TABLE from 'ipfs://$QMHASH' FILES = ('$FILE.csv', '${FILE}_v1.csv') FILE_FORMAT = (type = 'CSV' field_delimiter = ','  record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:26]: Remove this case.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  # copy dir with pattern
+  "copy into $TABLE from 'ipfs://$QMHASH' PATTERN = 'ontime.*csv' FILE_FORMAT = (type = 'CSV' field_delimiter = ',' record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:28]: Remove this case.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+  # copy parquet
+  "copy into $TABLE from 'ipfs://$QMHASH'  PATTERN = 'ontime.*parquet' FILE_FORMAT = (type = 'PARQUET')"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:30]: Remove this case.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 16]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../../../shell_env.sh
+TABLE=ontime200
+FILE=ontime_200
+QMHASH=QmPpCt1aYGb9JWJRmXRUnmJtVgeFFTJGzWFYEEX7bo9zGJ
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:7]: Visit https://ipfs.filebase.io/ipfs/QmPpCt1aYGb9JWJRmXRUnmJtVgeFFTJGzWFYEEX7bo9zGJ for files list.

Don't add tests that we don't have related files.

Comment (Xuanwo) [reply tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:7]: And use `https://ipfs.filebase.io` as endpoint which is faster than `https://ipfs.io` for our test cases.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 17]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../../../shell_env.sh
+TABLE=ontime200
+FILE=ontime_200
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:6]: Overdesign...

We are writing tests which won't change very often.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 18]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../../../shell_env.sh
+QMHASH=QmPpCt1aYGb9JWJRmXRUnmJtVgeFFTJGzWFYEEX7bo9zGJ
+
+echo "drop table if exists ontime_199;" | $MYSQL_CLIENT_CONNECT
+
+## Create table
+cat $CURDIR/../ddl/ontime.sql | sed "s/ontime/ontime_199/g" | $MYSQL_CLIENT_CONNECT
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:10]: We can use `ontime_200_from_ipfs` for table name.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+  "copy into ontime_200 from 'ipfs://$QMHASH/ontime.csv' CONNECTION = (ENDPOINT_URL='https://ipfs.filebase.io') FILE_FORMAT = (type = 'CSV' field_delimiter = ','  record_delimiter = '\n' skip_header = 1)"
+  # copy gzip csv
+  "copy into ontime_200 from 'ipfs://$QMHASH/ontime.csv.gz' CONNECTION = (ENDPOINT_URL='https://ipfs.filebase.io') FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'gzip'  record_delimiter = '\n' skip_header = 1)"
+  # copy zstd csv
+  "copy into ontime_200 from 'ipfs://$QMHASH/ontime.csv.zst' CONNECTION = (ENDPOINT_URL='https://ipfs.filebase.io') FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'zstd'  record_delimiter = '\n' skip_header = 1)"
+  # copy bz2 csv
+  "copy into ontime_200 from 'ipfs://$QMHASH/ontime.csv.bz2' CONNECTION = (ENDPOINT_URL='https://ipfs.filebase.io') FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'bz2'  record_delimiter = '\n' skip_header = 1)"
+  # copy xz csv
+  "copy into ontime_200 from 'ipfs://$QMHASH/ontime.csv.xz' CONNECTION = (ENDPOINT_URL='https://ipfs.filebase.io') FILE_FORMAT = (type = 'CSV' field_delimiter = ',' compression = 'xz'  record_delimiter = '\n' skip_header = 1)"
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_ipfs_location.sh:22]: Man, we don't have this file.

https://ipfs.filebase.io/ipfs/QmPpCt1aYGb9JWJRmXRUnmJtVgeFFTJGzWFYEEX7bo9zGJ

image

**Review by Xuanwo — APPROVED** [review]
Nice!

**Review by zhang2014 — APPROVED** [review]

**Comment by BohuTANG** [comment]
Nice.
@soyeric128 for documention 🚀

**Pull request merged**